### ISSUE TITLE
change node changes only first context entry in custom context

### ISSUE THREAD

**Issue by mw75** [node-red/node-red#4447]
### Current Behavior

I have a change note with three set rules:
SET flow.foo on custom to the value (string) "a"
SET flow.bar on custom to the value (string) "b"
SET global.foo on custom to the value (string) "c"

After triggering by an insert node the set method of the custom context module is only called for the first one.

### Expected Behavior

I expected all three context variables to be set by calling set method three times, once for each variable, or two times, once per scope. 

### Steps To Reproduce

create a contextStorage module with at least debug output on the methods. 
```js
...
set(scope, key, value, callback) {
        console.log("OdeonContextStore","set","scope",scope,"key",key,"value",value,"callback",typeof callback);
...
```

Import the attached flow, deploy and trigger the inject. Review your console for generated output. If you have a complete contextStorage module implementation review your context in the corresponding sidebar.

### Example flow

```
[
    {
        "id": "45e4e84a789d87e4",
        "type": "inject",
        "z": "3e854324c3d9dadd",
        "name": "",
        "props": [
            {
                "p": "payload"
            }
        ],
        "repeat": "",
        "crontab": "",
        "once": false,
        "onceDelay": 0.1,
        "topic": "",
        "payload": "",
        "payloadType": "date",
        "x": 520,
        "y": 400,
        "wires": [
            [
                "8e8960aa6d1f1180"
            ]
        ]
    },
    {
        "id": "8e8960aa6d1f1180",
        "type": "change",
        "z": "3e854324c3d9dadd",
        "name": "",
        "rules": [
            {
                "t": "set",
                "p": "#:(custom)::foo",
                "pt": "flow",
                "to": "a",
                "tot": "str",
                "dc": true
            },
            {
                "t": "set",
                "p": "#:(custom)::bar",
                "pt": "flow",
                "to": "b",
                "tot": "str"
            },
            {
                "t": "set",
                "p": "#:(custom)::foo",
                "pt": "global",
                "to": "c",
                "tot": "str"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 720,
        "y": 400,
        "wires": [
            [
                "425c993295afe887"
            ]
        ]
    },
    {
        "id": "425c993295afe887",
        "type": "debug",
        "z": "3e854324c3d9dadd",
        "name": "",
        "active": true,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "true",
        "targetType": "full",
        "statusVal": "",
        "statusType": "auto",
        "x": 890,
        "y": 400,
        "wires": []
    }
]
```

### Environment

- Node-RED version: 3.1.0
- Node.js version: 18.17.1
- npm version: 10.2.4
- Platform/OS: Linux
- Browser: Chrome

**Comment by hardillb** [comment]
I've added your debug to both the memory store and the localfs context that ships with and defined both in the settings.js and I'm not seeing this.

```
contextStorage: {
        default: {
            module: "memory"
        },
        custom: {
            module:"localfilesystem"
        },
    },
```

I do get a extra output when saving to the `custom` store because it uses a separate instance the memory context as a cache.

Are you sure your context is not throwing an error somewhere which is causing the code calling it to fail?

**Comment by mw75** [comment]
no, i don't.

**tldr**: discovered that i was missing the callback on the set method and that fixed the problem. Left all the text as documentation for others.

Just recapped the whole scenario by the following steps:
- removed $HOME/.node-red
- created a new directory "context-test" and changed to it
- run: npm i node-red
- run: ./node_modules/.bin/node-red and stopped it after startup completes to get a reference settings file
- just uncommented the contextStorage section and edited to:
```JS
      contextStorage: {
          default: {
              module:"memory"
          },
          custom: {
              module:require("./contextstub.js")
          }
      },
```
- added $HOME/.node-red/contextstub.js with
```JS
class context{
  constructor(config){console.log("constructor",arguments)};
  open(){console.log("open",arguments)};
  close(){console.log("close",arguments)};
  get(){console.log("get",arguments)};
  set(){console.log("set",arguments)};
  keys(){console.log("keys",arguments)};
  delete(){console.log("delete",arguments)};
  clean(){console.log("clean",arguments)};
}
module.exports=function(){
  return new context();
}
```
- run: ./node_modules/.bin/node-red
- opened the editor, import and deployed the flow above
- triggered inject and got the following on the console:

```TEXT
24 Nov 09:25:32 - [info] 

Welcome to Node-RED
===================

24 Nov 09:25:32 - [info] Node-RED version: v3.1.0
24 Nov 09:25:32 - [info] Node.js  version: v18.12.1
24 Nov 09:25:32 - [info] Linux 5.15.86-1-lts x64 LE
24 Nov 09:25:32 - [info] Loading palette nodes
24 Nov 09:25:33 - [info] Settings file  : /home/mwolff/.node-red/settings.js
24 Nov 09:25:33 - [info] Context store  : 'default' [module=memory]
constructor [Arguments] {}
24 Nov 09:25:33 - [info] Context store  : 'custom' [module=custom]
open [Arguments] {}
24 Nov 09:25:33 - [info] User directory : /home/mwolff/.node-red
24 Nov 09:25:33 - [warn] Projects disabled : editorTheme.projects.enabled=false
24 Nov 09:25:33 - [info] Flows file     : /home/mwolff/.node-red/flows.json
24 Nov 09:25:33 - [warn] 

---------------------------------------------------------------------
Your flow credentials file is encrypted using a system-generated key.

If the system-generated key is lost for any reason, your credentials
file will not be recoverable, you will have to delete it and re-enter
your credentials.

You should set your own key using the 'credentialSecret' option in
your settings file. Node-RED will then re-encrypt your credentials
file using your chosen key the next time you deploy a change.
---------------------------------------------------------------------

24 Nov 09:25:33 - [warn] Encrypted credentials not found
24 Nov 09:25:33 - [info] Server now running at http://127.0.0.1:1880/
24 Nov 09:25:33 - [info] Starting flows
24 Nov 09:25:33 - [info] Started flows
24 Nov 09:25:43 - [info] Stopping flows
24 Nov 09:25:43 - [info] Stopped flows
clean [Arguments] {
  '0': [
    '6a2eeeec0b6fe9ac',
    '45e4e84a789d87e4',
    '8e8960aa6d1f1180',
    '425c993295afe887'
  ]
}
24 Nov 09:25:43 - [info] Updated flows
24 Nov 09:25:43 - [info] Starting flows
24 Nov 09:25:43 - [info] Started flows
set [Arguments] {
  '0': '6a2eeeec0b6fe9ac',
  '1': 'foo',
  '2': 'a',
  '3': [Function: callback]
}
```
- reviewed https://nodered.org/docs/api/context/methods/#contextstoresetscope-key-value-callback and decided to modify my stub to
```JS
class context{
 ...
  set(scope, key, value,callback){
    console.log("set",arguments);
    if(typeof callback==="function"){
      callback(null)
    }
  };
...
}
module.exports=function(){
  return new context();
}
```
- rerun again and voila, it worked!
- just for information: there are three calls to set, one for each variable!

Sorry for missing that in the first place and taking your time!

Still wondering where and when delete method is called. See slack->core-dev also.

**Comment by hardillb** [comment]
You can delete items in context from the context sidebar view, try that

**Comment by mw75** [comment]
That calls set(key,undefined,callback). Signature of delete is delete(scope) not delete(key).
